sockeye/utils.py

```diff
@@ -271,49 +272,99 @@ def get_num_gpus() -> int:
 
 
 @contextmanager
-def acquire_gpu(lock_dir: str = "/var/lock", retry_wait: int = 10):
+def acquire_gpus(num_gpus:int, lock_dir: str = "/tmp", retry_wait_min: int = 30, retry_wait_rand: int = 30):
     """
-    Acquires a GPU by locking a file (therefore this assumes that everyone using GPUs calls this method and shares the
-    lock directory).
+    Acquire a number of GPUs in a transactional way. Will try to acquire all the requested number of GPUs. If currently
+    not enough GPUs are available all locks will be released and we wait until we retry. Will retry until enough
+    GPUs become available.
+
+    :param num_gpus: The number of GPUs to acquire.
+    :param lock_dir: The directory for storing the lock file.
+    :param retry_wait_min: The minimum number of seconds to wait between retries.
+    :param retry_wait_rand: Randomly add between 0 and `retry_wait_rand` seconds to the wait time.
+    :return: yields a list of GPU ids.
+    """
+    num_gpus_available = get_num_gpus()
+
+    if num_gpus > num_gpus_available:
+        raise ValueError("Requested %d GPUs, but only %d are available." % (num_gpus, num_gpus_available))
+
+    logger.info("Attempting to acquire %d GPUs of %d GPUs.", num_gpus, num_gpus_available)
+
+    while True:
+        with ExitStack() as exit_stack:
+            acquired_gpus = []
+            for _ in range(num_gpus):
+                gpu_id = exit_stack.enter_context(acquire_gpu(lock_dir))
+                if gpu_id is not None:
+                    acquired_gpus.append(gpu_id)
+                else:
+                    break
+            if len(acquired_gpus) == num_gpus:
+                try:
+                    yield acquired_gpus
+                finally:
+                    return
+        # couldn't acquire all GPUs, let's wait and try again later
+
+        # randomize so that multiple processes starting at the same time don't retry at a similar point in time
+        if retry_wait_rand > 0:
+            retry_wait_actual = retry_wait_min + random.randint(0, retry_wait_rand)
+        else:
+            retry_wait_actual = retry_wait_min
+        logger.info("Not enough GPUs available will try again in %ss." % retry_wait_actual)
+        time.sleep(retry_wait_actual)
+
+
+@contextmanager
+def acquire_gpu(lock_dir: str = "/var/lock"):
```

Comment:
This is now private to the module, right? (Single GPU should be acquired through acquire_gpus(1)). Rename to _acquire_gpu. Or even more _acquire_single_gpu to avoid confusion with the _gpus function.